### DISCUSSION TITLE
does not infer types on client

### DISCUSSION THREAD

**Discussion by af4oz** [trpc/trpc#1693] [General]
I am using pnpm workspaces, typescript 4.6, @trpc/* 9.20.3

Here's [code](https://github.com/0xafz/starter-fullstack-rest-trpc/tree/3e82345a00eae2ecd599354aac16bccf34726984)

It's able to infer types on server side,  but not on client.

**Comment by timcole** [comment]
Looking at your setup it seems that typescript has no way to know how to interact with code outside of the package's root (the app package in this case).

Because of this when importing `AppRouter` in `packages/app/src/utils/trpc.ts` typescript can’t process what things are (outside of the normal @trpc/server types since they’re installed on the client too) so you end up with the following type  (note the unknown generic).
```ts
Router<CreateContextOptions, unknown, any, any, any, DefaultErrorShape>
```

To combat this you can add a [reference](https://www.typescriptlang.org/docs/handbook/project-references.html) in your `packages/app/tsconfig.json` that way typescript knows where to interact with this outside code.
```
  "references": [
    { "path": "../api" }
  ]
```

After adding this reference the AppRouter in the app package should match the AppRouter in api package like so
```ts
Router<CreateContextOptions, CreateContextOptions, Record<"healthz", Procedure<CreateContextOptions, CreateContextOptions, undefined, undefined, string>> & Prefixer<...>, Prefixer<...>, Prefixer<...>, DefaultErrorShape>
```

**Reply by af4oz** [reply]
Thank you, that worked! I am also trying to replicate [KATT/zart](https://github.com/KATT/zART) setup (without references) with pnpm workspaces.

**Comment by uxduck** [comment]
_FWIW - Somewhat related to no types being inferred in the client:_
I had to restart VSCode quite a few times when first learning tRPC and integrating it with other things (Astro, Cloudflare pages/workers). If something doesn't work, I'd suggest first restarting VSCode.

**Reply by mrgoonie** [reply]
> 

True, but you don't have to restart VSCode, just restart `Typescript: Restart TS Server` should work (and faster).
